background/index.ts

```diff
@@ -94,6 +95,126 @@ async function main() {
     { pendingTransaction: {} }
   )
 
+  // Hold the current provider and block subscriber
+  const indexBalanceContext: {
+    provider?: JsonRpcProvider
+    blockSubscriber?: (blockNumber: number) => Promise<void>
+  } = {}
+
+  const indexBalanceOnBlock = async () => {
+    const { network, networkActive } = storage.get()
+    const { nodeRpcUrl, chainId } = network[networkActive]
+
+    // Set `{ staticNetwork: true }` to avoid infinite retries if nodeRpcUrl fails.
+    // Refer: https://github.com/ethers-io/ethers.js/issues/4377
+    indexBalanceContext.provider = new JsonRpcProvider(nodeRpcUrl, chainId, {
+      staticNetwork: true
+    })
```

Comment:
how about we double check if network is changed here?
```suggestion
    // Set `{ staticNetwork: true }` to avoid infinite retries if nodeRpcUrl fails.
    // Refer: https://github.com/ethers-io/ethers.js/issues/4377
    if (
      !indexBalanceContext.provider ||
      (await indexBalanceContext.provider.getNetwork()).chainId !==
        number.toBigInt(chainId)
  ) {
      indexBalanceContext.provider = new JsonRpcProvider(nodeRpcUrl, chainId, {
        staticNetwork: true
      })
    }
```

Reply:
Thank you very much for helping with the review!
Since we switched to using Storage's subscriber to ensure the network, there's no need to determine the current chainId.
Sorry for any inconvenience caused.